File: Python/linked_list/delete_node.py
"""
DATE: April 5, 2026
PATTERN: Linked List Traversal

APPROACHES:
1. BRUTE FORCE:
- Step 1: Rebuild list from array excluding last value.
- TIME: O(n) | SPACE: O(n)

2. BETTER:
- Step 1: Count length then remove node at position length-1.
- TIME: O(n) | SPACE: O(1)

3. OPTIMAL:
- Step 1: Walk until temp.next.next is None; set temp.next=None (delete tail).
- TIME: O(n) | SPACE: O(1)

AHA-MOMENT: Deleting last node only needs the penultimate node’s next pointer cleared.
"""

class Node:

    def __init__(self,val1,next1=None):
        self.val=val1
        self.next=next1

class Solution:
    # O(n),O(1)
    def delete_node(self,head):
        if head is None or head.next is None:
            return None

        temp=head
        while temp.next.next is not None:
            temp=temp.next;

        temp.next=None
        return head

    def printList(self,head):
        temp=head
        while temp:
            print(temp.val, end=" ")
            temp=temp.next

        print(end=" ")


if __name__=="__main__":
    sol=Solution()
    head=Node(1)
    head.next=Node(2)
    head.next.next=Node(3)
    head.next.next.next=Node(4)

    sol.printList(head)

    head=sol.delete_node(head)

    sol.printList(head)

    
        

File: C++/linked_list/delete_node.cpp
#include<bits/stdc++.h>
using namespace std;

class Node
{
    public:
    int val;
    Node* next;

    Node(int val1, Node* next1)
    {
        val=val1;
        next=next1;
    }
    Node(int val1)
    {
        val=val1;
        next=nullptr;
    }
};

class Solution
{
    public:
    Node* delete_node(Node *head)
    {
        if(head==nullptr || head->next==nullptr)
        {
            delete head;
            return nullptr;

        }
        Node * temp=head;
        while(temp->next->next!=nullptr)
        {
            temp=temp->next;
        }
        Node * drop= temp->next;
        temp->next=nullptr;
        delete drop;
        return head;

    }

    void printList(Node * head)
    {
        Node * temp=head;
        while(temp!=nullptr)
        {
            cout<<temp->val<<" ";
            temp=temp->next;

        }

    }
};

int main()
{
    Solution sol;
    Node * head=new Node(1);
    head->next=new Node(2);
    head->next->next=new Node(3);
    head->next->next->next=new Node(4);

    sol.printList(head);

    head=sol.delete_node(head);

    sol.printList(head);

}